File: Java/sorting/SelectionSort.java
/*
 * Selection Sort is a simple comparison-based sorting algorithm.
 * The idea is to repeatedly find the minimum element from the unsorted part
 * and place it at the beginning of the array.
 */

public class SelectionSort {
    public static void main(String[] args) {
        // Initialize an array of integers
        int[] array = { 4, 1, 9, 7, 2, 6, 5, 3, 8 };

        // Print the array before sorting
        System.out.print("Before Sorting: ");
        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }

        // Outer loop - iterate over each element of the array
        for (int i = 0; i < array.length - 1; i++) {
            int minIndex = i; // Assume the current element is the minimum

            // Inner loop - find the minimum element in the unsorted part
            for (int j = i + 1; j < array.length; j++) {
                if (array[j] < array[minIndex]) {
                    minIndex = j; // Update minIndex if a smaller element is found
                }
            }

            // Swap the found minimum element with the first element of the unsorted part
            int temp = array[minIndex];
            array[minIndex] = array[i];
            array[i] = temp;
        }

        // Print the array after sorting
        System.out.print("\nAfter Sorting: ");
        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }
    }
}

File: CPP/sorting/selection_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void selectionSort(vector<int> &arr, int n) // Function to perform selection sort
{
 for (int i = 0; i < n - 1; i++) // Loop through each element in the array
 {
  int minIndex = i; // Assume the current index is the minimum

  for (int j = i + 1; j < n; j++) // Loop to find the minimum element in the unsorted part
  {
   if (arr[j] < arr[minIndex]) // If a smaller element is found
   {
    minIndex = j; // Update the index of the minimum element
   }
  }

  if (minIndex != i) // If the minimum element is not at the current position
  {
   swap(arr[i], arr[minIndex]); // Swap the found minimum element with the first element of the unsorted part
  }
 }
}

void printArray(vector<int> &arr, int n) // Function to print the elements of the vector
{
 cout << "Sorted Array ";
 for (int i = 0; i < n; i++)
 {
  cout << arr[i] << " ";
 }
}

void inputArr(vector<int> &arr) // Function to input elements into the vector
{

 int n;
 cout << "Enter number of elements ";
 cin >> n;
 cout << "Enter elements ";

 for (int i = 0; i < n; i++) // Loop to read n elements
 {
  int x;
  cin >> x;
  arr.push_back(x);
 }
}

int main()
{
 vector<int> arr; // Declare a vector to hold the array elements

 inputArr(arr); // Input elements into the vector

 int n = arr.size(); // Get the size of the vector

 selectionSort(arr, n); // Call the selection sort function

 printArray(arr, n); // Print the sorted array
}
